packages/protocol/contracts/identity/FederatedAttestations.sol

```diff
@@ -0,0 +1,160 @@
+  ICeloVersionedContract,
+  Ownable,
+  Initializable,
+  UsingRegistry,
+  ReentrancyGuard,
+  UsingPrecompiles
+{
+  using SafeMath for uint256;
+  using SafeCast for uint256;
+
+  struct Attestation {
+    address account;
+    uint256 issuedOn;
+    address signer;
+  }
+  // identifier -> issuer -> attestations
+  mapping(bytes32 => mapping(address => Attestation[])) public identifierToAddresses;
+  // account -> issuer -> identifiers
+  mapping(address => mapping(address => bytes32[])) public addressToIdentifiers;
+  // signer => revocation time
+  mapping(address => uint256) public revokedSigners;
+
+  // TODO ASv2 Event declarations
+
+  /**
+   * @notice Sets initialized == true on implementation contracts
+   * @param test Set to true to skip implementation initialization
+   */
+  constructor(bool test) public Initializable(test) {}
+
+  /**
+   * @notice Used in place of the constructor to allow the contract to be upgradable via proxy.
+   * @param registryAddress The address of the registry core smart contract.
+   */
+  function initialize(address registryAddress) external initializer {
+    _transferOwnership(msg.sender);
+    setRegistry(registryAddress);
+    // TODO ASv2 initialize any other variables here
+  }
+
+  /**
+   * @notice Returns the storage, major, minor, and patch version of the contract.
+   * @return The storage, major, minor, and patch version of the contract.
+   */
+  function getVersionNumber() external pure returns (uint256, uint256, uint256, uint256) {
+    return (1, 1, 0, 0);
+  }
+
+  function _isRevoked(address signer, uint256 time) internal returns (bool) {
+    if (revokedSigners[signer] > 0 && revokedSigners[signer] >= time) {
+      return true;
+    }
+    return false;
+  }
+
+  function lookupAttestations(string identifier, address[] trustedIssuers)
+    public
+    view
+    returns (Attestation[])
+  {
+    Attestation[] memory attestations = new Attestation[];
+    for (uint256 i = 0; i < trustedIssuers.length; i++) {
+      address trustedIssuer = trustedIssuers[i];
+      for (uint256 j = 0; j < identifierToAddresses[identifier][trustedIssuer].length; j++) {
+        Attestation memory attestation = identifierToAddresses[identifier][trustedIssuer][j];
+        if (!_isRevoked(attestation.signer, attestation.issuedOn)) {
+          attestations.push(attestation);
+        }
+      }
+    }
+    return attestations;
+  }
+
+  function lookupIdentifiersbyAddress(address account, address[] trustedIssuers)
+    public
+    view
+    returns (bytes32[])
+  {
+    bytes32[] memory identifiers = new bytes32[];
+    for (uint256 i = 0; i < trustedIssuers.length; i++) {
+      address trustedIssuer = trustedIssuers[i];
+      for (uint256 j = 0; j < addressToIdentifiers[account][trustedIssuer].length; j++) {
+        bytes32 memory identifier = addressToIdentifiers[account][trustedIssuer][j];
+        Attestation memory attestation = identifierToAddresses[identifier][trustedIssuer];
```

Comment:
hmm running into some issues with this, since this is a list of attestations, not a single one. How do we want to figure out which attestation to choose?
- we could allow for duplicate identifiers in the `addressToIdentifiers` mapping, and require that the indices of identifiers & attestations stay in sync; this would need to be accounted for in the delete function
- we can iterate through the attestations and take the most recent, non-revoked attestation
Other ideas? But in its current form I think this is causing some issues throughout; may help to re-tackle this once we have some unit tests in place

Reply:
whatever we do choose, we should be really explicit about what invariants we have laid out and comment this in these functions